clang/lib/CodeGen/CodeGenModule.cpp

```diff
@@ -6853,6 +6853,13 @@ void CodeGenModule::EmitTopLevelDecl(Decl *D) {
         if (ES->hasExternalDefinitions(D) == ExternalASTSource::EK_Never)
           DI->completeUnusedClass(*CRD);
     }
+    // If we're emitting a dynamic class from the importable module we're
+    // emitting, we always need to emit the virtual table according to the ABI
+    // requirement.
+    if (CRD->getDefinition() && CRD->isDynamicClass() &&
+        CRD->isInCurrentModuleUnit())
+      EmitVTable(CRD);
```

Comment:
Yes, this can happen with the two phase compilation of modules (https://clang.llvm.org/docs/StandardCPlusPlusModules.html#how-to-produce-a-bmi). Which will compile a `.cppm` file to a `.pcm` file, and we will compile the `.pcm` file the `.o` file.

The one phase compilation can also happen for modules.

Reply:
You mean Sema::DefineUsedVTables doesn't run at all in such cases?

My primary concern here is that having multiple overlapping codepaths is hard to maintain.  If Sema::DefineUsedVTables can't actually force the definition in all the relevant cases, we should just eliminate HandleVTable.

Reply:
`Sema::DefineUsedVTables` is executed in all cases. However, in two phase compilation, the consumer in the first phase is not the CodeGenConsumer but the ASTWriter, so when the code get executed: https://github.com/llvm/llvm-project/blob/876c6204f12fa2738ff8ca886e664b826847d6d4/clang/lib/Sema/SemaDeclCXX.cpp#L18758-L18759

the corresponding `HandleVTable` in CodeGen Consumer won't get executed.

In the second phase, the CodeGen consumer will be the consumer, and the AST Reader will pass the interesting decls to the consumer: https://github.com/llvm/llvm-project/blob/e9a362362e03c6aea7a64473e8b8bb9140fd3855/clang/lib/Serialization/ASTReader.cpp#L8052-L8057

Reply:
So DefineUsedVTables is called, but the ASTWriter eats the call without actually recording the fact anywhere?  Okay, that makes more sense.  So I guess that gives a potential alternate approach; we could teach ASTWriter to record the fact that HandleVTable was called, and replay it later.

But if we can reliably handle emitting vtables in EmitTopLevelDecl, it's probably simpler to just eliminate HandleVTable.

Reply:
> So DefineUsedVTables is called, but the ASTWriter eats the call without actually recording the fact anywhere? Okay, that makes more sense. So I guess that gives a potential alternate approach; we could teach ASTWriter to record the fact that HandleVTable was called, and replay it later.

Done. Although it is somewhat not efficient, I made it if it makes the structure more clear.

> But if we can reliably handle emitting vtables in EmitTopLevelDecl, it's probably simpler to just eliminate HandleVTable.

I tried but it looks not so easy. The problem is that `EmitTopLevelDecl` getting called on the fly during the parsing but `Sema::DefineUsedVTables` is only get called after Sema finished parsing the translation unit. So we still need some special function to replace `HandleVTable`, but then it is meaningless.

Reply:
The new code makes sense to me.